File: SY/graph/1991.py
class Node:
	def __init__(self, data, left, right):
		self.data=data
		self.left=left
		self.right=right

n=int(input())
tree={}

for i in range(n):
	data, left, right=input().split()
	if left=='.':
		left=None
	if right=='.':
		right=None
	tree[data]=Node(data,left,right)

#전위 순회: 루트 -> 왼쪽 자식 -> 오른쪽 자식
def pre_order(node):
	print(node.data,end='')
	if node.left!=None:
		pre_order(tree[node.left])
	if node.right!=None:
		pre_order(tree[node.right])

#중위 순회: 왼쪽 자식 -> 루트 -> 오른쪽 자식
def in_order(node):
	if node.left!=None:
		in_order(tree[node.left])
	print(node.data,end='')
	if node.right!=None:
		in_order(tree[node.right])

#후위 순회: 왼쪽 자식 -> 오른쪽 자식 -> 루트
def post_order(node):
	if node.left!=None:
		post_order(tree[node.left])
	if node.right!=None:
		post_order(tree[node.right])
	print(node.data,end='')

pre_order(tree['A'])
print()
in_order(tree['A'])
print()
post_order(tree['A'])

File: JH/DAY29/1991.cpp
#import <bits/stdc++.h>

using namespace std;



int n;

pair<char,char>arr[26];


void preorder(char x){
    if(x!='.'){
        cout << x;
        preorder(arr[x-'A'].first);
        preorder(arr[x-'A'].second);

    }
}

void inorder(char x){
    if(x!='.'){
        inorder(arr[x-'A'].first);
        cout << x ;
        inorder(arr[x-'A'].second);
    }
}

void postorder(char x){
    if(x!='.'){
        postorder(arr[x-'A'].first);
        postorder(arr[x-'A'].second);
        cout << x ;
    }
}



int main(){

    cin >> n;

    for(int i = 0 ; i < n ; i++){
        char node, left, right;
        cin >> node >> left >> right;
        arr[node-'A'].first = left;
        arr[node-'A'].second = right;
    }

    preorder('A');
    cout << "\n";
    inorder('A');
    cout << "\n";
    postorder('A');
    cout << "\n";












}